capreolus/searcher/special.py

```diff
@@ -0,0 +1,116 @@
+        runs = self.convert_to_trec_runs(extract_dir / extract_file_name, train=True)
+        return runs
+
+
+@Searcher.register
+class MsmarcoPsg(Searcher, MsmarcoPsgSearcherMixin):
+    module_name = "msmarcopsg"
+    dependencies = [Dependency(key="benchmark", module="benchmark", name="msmarcopsg")]
+
+    def _query_from_file(self, topicsfn, output_path, cfg):
+        """ only query results in dev and test set are saved """
+        final_runfn = output_path / "searcher"
+        final_donefn = output_path / "done"
+        if os.path.exists(final_donefn):
+            return output_path
+
+        tmp_dir = self.get_cache_path() / "tmp"
+        tmp_dir.mkdir(exist_ok=True, parents=True)
+        output_path.mkdir(exist_ok=True, parents=True)
+
+        # train
+        train_run = self.download_and_prepare_train_set(tmp_dir=tmp_dir)
+        self.write_trec_run(preds=train_run, outfn=final_runfn, mode="wt")
+
+        # dev and test
+        dev_test_urls = [
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.dev.tar.gz",
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.eval.tar.gz",
+        ]
+        runs = {}
+        for url in dev_test_urls:
+            extract_file_name = self.get_fn_from_url(url)
+            extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+            runs.update(self.convert_to_trec_runs(extract_dir / extract_file_name, train=False))
+        self.write_trec_run(preds=runs, outfn=final_runfn, mode="a")
+
+        with open(final_donefn, "wt") as f:
+            print("done", file=f)
+        return output_path
+
+
+@Searcher.register
+class MsmarcoPsgBm25(BM25, MsmarcoPsgSearcherMixin):
+    module_name = "msmarcopsgbm25"
+    dependencies = [
+        Dependency(key="benchmark", module="benchmark", name="msmarcopsg"),
+        Dependency(key="index", module="index", name="anserini"),
+    ]
+
+    def _query_from_file(self, topicsfn, output_path, config):
+        final_runfn = output_path / "searcher"
+        final_donefn = output_path / "done"
+        if final_donefn.exists():
+            return output_path
+
+        output_path.mkdir(exist_ok=True, parents=True)
+        tmp_dir = self.get_cache_path() / "tmp"
+        tmp_topicsfn = tmp_dir / os.path.basename(topicsfn)
+        tmp_output_dir = tmp_dir / "BM25_results"
+        tmp_output_dir.mkdir(exist_ok=True, parents=True)
+
+        train_runs = self.download_and_prepare_train_set(tmp_dir=tmp_dir)
+        with open(tmp_topicsfn, "wt") as f:
+            for qid, title in tqdm(load_trec_topics(topicsfn)["title"].items()):
+                if qid not in self.benchmark.folds["s1"]["train_qids"]:
+                    f.write(topic_to_trectxt(qid, title))
+        super()._query_from_file(topicsfn=tmp_topicsfn, output_path=tmp_output_dir, config=config)
```

Comment:
This line would trigger an anserini index creation on MSMarco right? Do you know how long that took (approximately) for you? Just curious :)

Reply:
not exactly sure but within 1 or 2 hours I think? will update u when i tried it again ;)